tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java

```diff
@@ -0,0 +1,683 @@
+   * @return the mean squared logarithmic percentage error
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> meanSquaredLogarithmicError(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = tf.dtypes.cast(labels, predictions.asOutput().dataType());
+    Tuple<T> ops = LossesImpl.squeezeOrExpandDimensions(tf, tLabels, predictions, null);
+    predictions = ops.getTarget();
+    tLabels = ops.getLabels();
+
+    Operand<T> epsilonConst = tf.dtypes.cast(tf.constant(EPSILON), dataType);
+    Operand<T> one = tf.dtypes.cast(tf.constant(1), dataType);
+
+    Operand<T> firstLog = tf.math.log(tf.math.add(tf.math.maximum(predictions, epsilonConst), one));
+    Operand<T> secondLog = tf.math.log(tf.math.add(tf.math.maximum(tLabels, epsilonConst), one));
+
+    return tf.math.mean(tf.math.squaredDifference(firstLog, secondLog), tf.constant(-1));
+  }
+
+  /**
+   * Computes the binary crossentropy loss between labels and predictions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets
+   * @param predictions the predictions
+   * @param fromLogits Whether to interpret predictions as a tensor of logit values
+   * @param labelSmoothing A number in the range [0, 1]. When 0, no smoothing occurs. When &gt; 0,
+   *     compute the loss between the predicted labels and a smoothed version of the true labels,
+   *     where the smoothing squeezes the labels towards 0.5. Larger values of labelSmoothing
+   *     correspond to heavier smoothing.
+   * @param <T> the data type of the predictions and labels
+   * @return the binary crossentropy loss.
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> binaryCrossentropy(
+      Ops tf, Operand<U> labels, Operand<T> predictions, boolean fromLogits, float labelSmoothing) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = tf.dtypes.cast(labels, dataType);
+    Tuple<T> ops = LossesImpl.squeezeOrExpandDimensions(tf, tLabels, predictions, null);
+    predictions = ops.getTarget();
+    tLabels = ops.getLabels();
+
+    if (labelSmoothing != 0.0f) {
+      tLabels = smoothLabelsBinaryX(tf, tLabels, labelSmoothing);
+    }
+    Operand<T> bce = binaryCrossentropy(tf, tLabels, predictions, fromLogits);
+    return tf.math.mean(bce, tf.constant(-1));
+  }
+
+  /**
+   * Compute binary crossentropy loss between labels and predictions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param target the target Operand
+   * @param output the output, either logits or a probability distribution
+   * @param fromLogits whether `output` is expected to be a logits tensor. By default, we consider
+   *     that `output` encodes a probability distribution.
+   * @param <T> the data type of the Operands
+   * @return the binary crossentropy loss.
+   */
+  private static <T extends TNumber> Operand<T> binaryCrossentropy(
```

Comment:
I tripped over this private method having the usual naming of a loss method, since I didn't notice that it was private and so expected it to follow the conventions of public loss methods, such as invoking `squeezeOrExpandDimensions`.  Also (if I'm navigating through unfamiliar territory accurately), this method doesn't compute a `binaryCrossentropy` since it is depending on its caller to compute the mean at the end.

Reply:
This method does the grunt work for the binaryCrossentropy after the operands have had their shapes and types manipulated and after smoothing the labels. Perhaps a new name would remove some of the confusion.

Reply:
Yes, I wonder if we want to call it something like `binaryCrossentropyHelper`?